[Teee728] iP

Teeesa7 opened the pull request:
### Tango the Chatbot

> "All Power is within You; <br> You can do anything and everything." - Swami Vivekananda - [source](https://www.goodreads.com/quotes/161483-all-power-is-within-you-you-can-do-anything-and)

Forgot what to do? No worries! Tango is here to save you! It's
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ fast to use

All you need to do is,
1. Download the JAR file from [here](https://github.com/Teee728/ip/releases/tag/A-Jar)
2. Follow the instructions given there
3. Once you get it running, add your tasks away! :star_struck:

And, it is **FREE!**

Features:
- [ ] Managing tasks
- [ ] Managing deadlines _(coming soon)_
- [ ] Reminders _(coming soon)_

---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Teeesa7 committed:
Added simple greet and exit messages.

Teeesa7 committed:
Added echo commands

Teeesa7 committed:
Adding and viewing task list feature created

Teeesa7 committed:
created a task class, removed indentation

Teeesa7 committed:
added classes for different types of tasks

Teeesa7 committed:
added input and expected output files

Teeesa7 committed:
removed original input file, replaced with "in"

Teeesa7 committed:
Added exception handling

Teeesa7 committed:
Updated expected text for new input

Teeesa7 committed:
Updated input cases to deal with exception handling

Teeesa7 committed:
Added "delete" command, fixed how Event reads duration from input

Teeesa7 committed:
Added enums for event types

Teeesa7 committed:
added methods to load and save tasks automatically

Teeesa7 committed:
Added method to identify dates and times with backward string compatibility, modified Deadline class

Teeesa7 committed:
Added new classes to make it more OOP.

Teeesa7 committed:
Organize project into packages

- Move classes into separate packages:
  - `duke.command`
  - `duke.exception`
  - `duke.task`
  - `duke.ui`
  - `duke.storage`

- Update import statements in affected classes

Teeesa7 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Teeesa7 committed:
Add JUnit tests for Parser, Storage, ToDo and TaskList

Teeesa7 committed:
no message

Teeesa7 committed:
Add JUnit tests and JavaDoc comments

Teeesa7 committed:
Fix coding standard

Teeesa7 committed:
Fix UI class

Teeesa7 committed:
Fix coding standard

Teeesa7 committed:
Add find method, fix backward string compatibility for Deadline and Event

Teeesa7 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/Parser.java
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskType.java

Teeesa7 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/Parser.java
#	src/main/java/duke/storage/Storage.java
#	src/main/java/duke/task/Deadline.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/ui/Ui.java

nichee commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,182 @@
+                    if (command.length() <= 5) {
+                        ui.showError("You gotta enter some task TO DO!");
+                    } else {
+                        tasks.addToDoTask(new ToDo(command.substring(5).trim()));
```
Consider replacing any "magic numbers" (e.g., 5 and 3 in substring(5) and parts[3]) with named constants or variables with meaningful names to improve code readability.

nichee commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,130 @@
+     *
+     * @param task The deleted task.
+     */
+    public static void printDeletedTaskMessage(Task task) {
```
Consider using String.format for better readability! No biggie though

nichee reviewed (commented):
LGTM, couldn't really find coding standard violations so a lil nitpicky.

Teeesa7 committed:
Add CheckStyle, fix JDK version

Teeesa7 committed:
Add GUI to the chatbot

Teeesa7 committed:
Fix UI

Teeesa7 committed:
Revert back to original UI

gowthaman-01 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+        for (int i = 0; i < taskList.size(); i++) {
+            Task task = taskList.getTask(i);
+             taskDetails += (i + 1) + ".[" + task.type + "][" + (task != null ? task.getStatusIcon() : "") + "] " + task.description;
+            if (task instanceof Deadline) {
```
Try to avoid instanceof, instead use polymorphism to achieve the same functionality

gowthaman-01 commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,177 @@
+    public TaskList() {
+        this.tasks = new ArrayList < > ();
+    }
+    /**
```
Add an empty line between methods

gowthaman-01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,82 @@
+     * Updates the status icon based on the completion status.
+     */
+    private void updateStatusIcon() {
+        this.statusIcon = (isDone ? "X" : " ");
```
Try to define "X" as a constant static variable

gowthaman-01 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,131 @@
+            }
+        }
+
+        try (FileWriter fileWriter = new FileWriter(file, false)) {
```
Avoid arrowhead style code!

gowthaman-01 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,201 @@
+
+        try {
+            ui.printDashes();
+            switch (words[0].toLowerCase()) {
```
Please double check the indentation of your switch statements

gowthaman-01 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,201 @@
+        try {
+            ui.printDashes();
+            switch (words[0].toLowerCase()) {
+                case "bye":
```
Try to use a enum instead of string matching

Teeesa7 committed:
Fix UI

Teeesa7 committed:
Delete duke.txt

Teeesa7 committed:
Add Java assertions

Java assertions are added to help identify and catch bugs early in the
development process.

* Task class: Ensure that task type is not null and necessary fields are
  initialized correctly.

* Deadline class: Ensure either LocalDateTime or byString is non-null in
  Deadline creation and handle missing task description.

* Event class: Ensure both start and end times are correctly parsed and
  handle missing task description.

* TaskList class: Ensure correct handling of task addition, deletion,
  and marking status.

* ToDo class: Ensure proper initialization and message generation for
  ToDo tasks.

* Ui class: Ensure appropriate error messages and responses are handled
  in the Ui class.

* Duke class: Ensure correct initialization of storage and task list
  in Duke class.

* Parser class: Ensure response is not null and running status is a
  boolean.

Teeesa7 committed:
Enhance Code Quality for Improved Readability and Maintainability

The code exhibits certain aspects that compromise its readability and
maintainability, and it deviates from Checkstyle guidelines.

To enhance the overall quality of the codebase, improve readability,
and make it align with Checkstyle guidelines for better consistency
and compliance.

* Helper functions are being extracted to condense lengthy methods,
  contributing to improved code organization and readability.
* Magic numbers are being replaced with named constants to enhance code
  clarity and facilitate easier maintenance.
* Guard clauses are being implemented at method entry points to create
  a cleaner and more understandable happy path.
* An enumeration for task status is being introduced to enhance code
  organization and ensure consistent handling of task statuses.

These changes aim to create a codebase that is not only more readable
and maintainable but also aligned with industry best practices and
coding standards.

Teeesa7 committed:
Enhancing Code Quality for Improved Readability and Maintainability

The code exhibits certain aspects that compromise its readability and
maintainability, and it deviates from Checkstyle guidelines.

To enhance the overall quality of the codebase, improve readability,
and make it align with Checkstyle guidelines for better consistency
and compliance.

* Helper functions are being extracted to condense lengthy methods,
  contributing to improved code organization and readability.
* Magic numbers are being replaced with named constants to enhance code
  clarity and facilitate easier maintenance.
* Guard clauses are being implemented at method entry points to create
  a cleaner and more understandable happy path.
* An enumeration for task status is being introduced to enhance code
  organization and ensure consistent handling of task statuses.

These changes aim to create a codebase that is not only more readable
and maintainable but also aligned with industry best practices and
coding standards.

Teeesa7 committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/Teee728/ip into branch-A-CodeQuality

* 'branch-A-CodeQuality' of https://github.com/Teee728/ip:
  Enhancing Code Quality for Improved Readability and Maintainability

Teeesa7 committed:
Merge pull request #1 from Teee728/branch-A-Assertions

Add Java assertions

Teeesa7 committed:
Merge pull request #2 from Teee728/branch-A-CodeQuality

Branch a code quality

Teeesa7 committed:
Add Prioritization feature

The task management system lacks a prioritization feature,
limiting users' ability to distinguish and prioritize tasks effectively.

Prioritizing tasks is essential for effective time management and
ensuring that critical tasks are addressed promptly. This feature will
enhance the overall usability of the system.

Introduce a priority tagging system using the "/p" tag, allowing users
to assign low, medium, or high priority to tasks. This empowers users to
organize and manage their tasks based on priority levels.

The "/p" tag provides a clear and intuitive method for users to indicate
the priority of their tasks. This approach is simple, consistent, and
aligns with user expectations.

Additionally, code refactoring has been performed to enhance readability
and comply with checkstyle guidelines. This ensures that the codebase
remains maintainable and follows best practices.

Teeesa7 committed:
Create gradle.yml

Teeesa7 committed:
Update gradle.yml

Teeesa7 committed:
Update gradle.yml

Teeesa7 committed:
Update gradle.yml

Teeesa7 committed:
Update gradle.yml

Teeesa7 committed:
Fix checkstyle errors for import statements

Teeesa7 committed:
Merge branch 'master' of https://github.com/Teee728/ip

Teeesa7 committed:
Fix Javadocs

Teeesa7 committed:
Add Background Image, Change profile pictures

Teeesa7 committed:
Make chatbot more personalized

- Change responses
- Change Package and Class names from "duke" to "cruisey"

Teeesa7 committed:
Fix error handling

- Delete command sends error message for wrong task number input
- Application exits after 2 second delay

Teeesa7 committed:
Delete data.txt

Teeesa7 committed:
Fix more error handling
- Correct name from "DukeException" to "CruiseyException"
- Fix time format for Deadline and Event classes
- Fix GUI font family

Teeesa7 committed:
Fix testing files for Storage, Tasklist, Todo and Parser

Teeesa7 committed:
Comment out ToDoTest

ToDoTest was commented out as it only works for CLI

Teeesa7 committed:
Merge branch 'branch-A-MoreErrorHandling'

# Conflicts:
#	data/duke.txt

Teeesa7 committed:
Merge branch 'branch-A-MoreTesting'

Teeesa7 committed:
Set chatbot title on application window

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Improve GUI by adding a HBox for the messages

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Add image of GUI

Teeesa7 committed:
Merge branch 'master' of https://github.com/Teee728/ip

Teeesa7 committed:
Fix Load() method for tasks with priority

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md

Teeesa7 committed:
Update README.md
